### PR TITLE
TYP: index.pyi

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -559,6 +559,7 @@ def from_tuples(
             if isinstance(tuples, Index):
                 tuples = tuples._values
 
+            tuples = cast(np.ndarray, tuples)
```
Comment (simonjayhawkins) [inline pandas/core/indexes/multi.py:562]: can you add a comment why we can do this for future readers. or a TODO if easy to explain how to rectify.

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -1111,8 +1112,8 @@ def _engine(self):
         # Check the total number of bits needed for our representation:
         if lev_bits[0] > 64:
             # The levels would overflow a 64 bit uint - use Python integers:
-            return MultiIndexPyIntEngine(self.levels, self.codes, offsets)
-        return MultiIndexUIntEngine(self.levels, self.codes, offsets)
+            return MultiIndexPyIntEngine(list(self.levels), self.codes, offsets)
```
Comment (simonjayhawkins) [inline pandas/core/indexes/multi.py:1115]: I don't think should need to change this. should probably be using sequence in BaseMultiIndexCodesEngine.\_\_init__

from https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#conventions

> avoid invariant collection types (list, dict) in argument positions, in favor of covariant types like Mapping or Sequence;

File: pandas/core/strings/object_array.py [hunk 3]
```diff
@@ -73,8 +73,9 @@ def _str_map(self, f, na_value=None, dtype: Optional[Dtype] = None):
             arr = np.asarray(arr, dtype=object)  # type: ignore[assignment]
         mask = isna(arr)
         convert = not np.all(mask)
+        ndarr = np.asarray(arr)
```
Comment (simonjayhawkins) [inline pandas/core/strings/object_array.py:76]: why is this added. not already done in L70-73?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @jbrockmendel

File: pandas/core/strings/object_array.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/strings/object_array.py:76]: good catch, can rename variable there (i think this might have snuck in from another branch as it appears unrelated to index.pyx)

**Comment by jbrockmendel** [comment]
Updated+green

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel for the updates

File: pandas/core/indexes/datetimelike.py [hunk 4]
```diff
@@ -536,7 +537,7 @@ def shift(self: _T, periods: int = 1, freq=None) -> _T:
     # --------------------------------------------------------------------
     # List-like Methods
 
-    def _get_delete_freq(self, loc: int):
+    def _get_delete_freq(self, loc: Union[int, Sequence[int]]):
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py:540]: could be a slice too (do we have a typing alias for Indexers?)

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, pls rebase

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: > (do we have a typing alias for Indexers?)

we do not.  i looked at implementing one a few weeks ago, dont remember off the top of my head why i stuck a pin in it

**Review by jreback — APPROVED** [review]
